theseus/src/data/profiles.rs

```diff
@@ -0,0 +1,492 @@
+use super::DataError;
+use crate::launcher::ModLoader;
+use daedalus::{minecraft::JavaVersion, modded::LoaderVersion};
+use futures::*;
+use once_cell::sync::OnceCell;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{HashMap, HashSet},
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use tokio::{
+    fs,
+    process::{Child, Command},
+    sync::{Mutex, RwLock, RwLockReadGuard},
+};
+
+static PROFILES: OnceCell<RwLock<Profiles>> = OnceCell::new();
+pub const PROFILE_JSON_PATH: &str = "profile.json";
+
+#[derive(Debug)]
+pub struct Profiles(pub HashMap<PathBuf, Profile>);
+
+// TODO: possibly add defaults to some of these values
+pub const CURRENT_FORMAT_VERSION: u32 = 1;
+pub const SUPPORTED_ICON_FORMATS: &[&'static str] = &[
+    "bmp", "gif", "jpeg", "jpg", "jpe", "png", "svg", "svgz", "webp", "rgb",
+    "mp4",
+];
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct Profile {
+    #[serde(skip)]
+    pub path: PathBuf,
+    pub metadata: Metadata,
+    pub java: JavaSettings,
+    pub memory: MemorySettings,
+    pub resolution: WindowSize,
+    pub hooks: ProfileHooks,
```

Comment:
These should all be optional, mainly so a checkbox in the UI can be created to "override" the values. If not, they should take the global settings values

Reply:
They do take those values at creation time in the initialization method (this prevents them from breaking). Fields are then overridden using builder methods.

Reply:
I don't think that should be the case. Profile settings should always sync with the global launcher settings. This is what other launchers like MultiMC do and make the most sense to the user

Reply:
Sorry, I didn't mean to include metadata + path in my selection. I think all the others should be optionals, including Java version.

It would be weird if in order to edit your global java flags, you have to go to each instance and manually edit it. If exceptions exist, then that is why people use overrides.

Reply:
Sorry, I didn't mean to include the metadata + path.

All the others should be options, so they can be in sync with the global settings unless needed to be overridden. This includes Java settings. It would be weird that in order to change your Java flags globally, you would have to go to each instance and manually edit it. If the global flags aren't compatible, then that is why the overrides exist. Users advanced enough to do custom java flags should know this